compiler/damlc/daml-stdlib-src/DA/Internal/Template.daml

```diff
@@ -50,6 +51,14 @@ stakeholder t = signatory t ++ observer t
 class Template t => Choice t c r | t c -> r where
   -- | Exercise a choice on the contract with the given contract ID.
   exercise : ContractId t -> c -> Update r
+  _toAnyChoice : proxy t -> c -> AnyChoice
+  _fromAnyChoice : proxy t -> AnyChoice -> Optional c
```

Comment:
I suppose these two methods have the `proxy t` and not the signatures from below because their signatures need to refer to `t`. Is that correct?

Reply:
There are two issues here:

1. Declaring the typeclass `IouInstance` will require `AllowAmbigousTypes` in every module that defines a template if we don’t have the proxy there which is obviously not a good idea (there is a comment in `ghc-lib` on this).
2. 1 applies to the declaration of the typeclass but not the declaration of the `Choice` instance. However, if we don’t have the proxy here we need to make it up during template desugaring to define `_toAnyChoice = _toAnyChoiceProposalArchive someProxy` and apart from that complicating template desugaring, I’m not really sure how to do this nicely without requiring `ScopedTypeVariables` in every module that defines a `template`.

I’ll add a comment for `2` here.

Reply:
Don't we have `ScopedTypeVariables` enabled by default in DAML?

Reply:
oh good point, I thought I had grepped for that but apparently I missed this.